### PR TITLE
Fix a couple goroutine leaks

### PR DESCRIPTION
**Pull request by gsalgado**
We were leaking up to two goroutines for every node started, and when
using loglevel>5, one of them would leave open FDs behind as well,
which would eventually cause the hive node to not be able to start
any more nodes due to having too many open files.

Also remove some trailing whitespace and simplify a couple things.

### CODE REVIEW

File: simulator.go [hunk 1]
```diff
@@ -266,14 +266,26 @@ func (h *simulatorAPIHandler) ServeHTTP(w http.ResponseWriter, r *http.Request)
 				http.Error(w, err.Error(), http.StatusInternalServerError)
 				return
 			}
-			logger = logger.New("id", container.ID[:8])
+			cLogger := logger.New("id", container.ID[:8])
```
Comment (karalabe) [inline simulator.go]: Why did you rename this? It will cause the loggers below not to have the `id` embedded any more. This will especially be bad for the new termination logs you added, since they don't have any context, so you'll just get a batch of `client container finished cleanly` messages one after the other. Am I understanding the code incorrectly?

File: containers.go [hunk 2]
```diff
@@ -361,5 +369,28 @@ func runContainer(daemon *docker.Client, id string, logger log15.Logger, logfile
 		logger.Error("failed to start container", "error", err)
 		return nil, err
 	}
+	waiter = fdClosingWaiter{
+		CloseWaiter: waiter,
+		closers:     fdsToClose,
+		logger:      logger,
+	}
```
Comment (karalabe) [inline containers.go]: Please refactor into 

```go
return fdClosingWaiter{
	CloseWaiter: waiter,
	closers:     fdsToClose,
	logger:      logger,
}, nil
```

I'd really like to avoid reassigning waiter with a completely different type that happens to implement the same interface.

File: containers.go [hunk 3]
```diff
@@ -312,22 +312,30 @@ func runContainer(daemon *docker.Client, id string, logger log15.Logger, logfile
 	// If we're the outer shell, log straight to stderr, nothing fancy
 	stdout := io.Writer(os.Stdout)
 	stream := io.Writer(os.Stderr)
+	var log, hookedR, hookedW *os.File
+	var fdsToClose []io.Closer
+	var err error
```
Comment (karalabe) [inline containers.go]: Declaring `err` here just makes the code convoluted. It "hints" that errors are propagated through for some reason. However the code below just checks for error conditions and either returns or continues. There's no need to declare it up front.

File: simulator.go [hunk 4]
```diff
@@ -266,14 +266,26 @@ func (h *simulatorAPIHandler) ServeHTTP(w http.ResponseWriter, r *http.Request)
...
 
 			logfile := fmt.Sprintf("client-%s.log", container.ID[:8])
-			if _, err = runContainer(h.daemon, container.ID, logger, filepath.Join(h.logdir, logfile), false); err != nil {
+			waiter, err := runContainer(h.daemon, container.ID, cLogger, filepath.Join(h.logdir, logfile), false)
+			if err != nil {
 				logger.Error("failed to start client", "error", err)
 				http.Error(w, err.Error(), http.StatusInternalServerError)
 				return
 			}
+			go func() {
```
Comment (karalabe) [inline simulator.go]: Please don't start a new goroutine here. Instead, add a field to `simulatorAPIHandler` https://github.com/karalabe/hive/blob/master/simulator.go#L192 below `nodes` called `waiters`, that you can iterate over after closing the containers https://github.com/karalabe/hive/blob/master/simulator.go#L368. 

You'd also need to add the wait logic to the node `DELETE` handler https://github.com/karalabe/hive/blob/master/simulator.go#L340.

Having the extra field is cleaner that one more concurrent goroutine which may run at an arbitrary point in time.

File: containers.go (hunk 3 above)
Comment (karalabe) [inline containers.go]: Actually, if I read the code correctly, non of these pre-declares (apart from the closer slice) you added are needed here, and they just convolute the code. Please revert to the original code that declared them when used.

File: simulator.go (hunk 1 above)
Comment (gsalgado) [inline simulator.go]: I thought we'd only want the client prefix in the log messages generated within runContainer(), but as you pointed out they're helpful for the termination logs I added, and probably also for other logs emitted further down, so I'll revert

File: containers.go (hunk 2 above)
Comment (gsalgado) [inline containers.go]: Ok, done

File: containers.go (hunk 3 above)
Comment (gsalgado) [inline containers.go]: The hookedW/hookedR/err declarations are leftovers from a previous version of this fix -- they're indeed not necessary, so I reverted them.

File: simulator.go (hunk 4 above)
Comment (gsalgado) [inline simulator.go]: That means we'll end up with a ton of idle goroutines (up to two for every node we start, depending on the loglevel used) lingering around until the simulator node is terminated, and since we'd only close them at the end, the log files would remain open until then and we'd probably still hit the open file limit and crash, which is what I'm trying to fix.

> You'd also need to add the wait logic to the node DELETE handler https://github.com/karalabe/hive/blob/master/simulator.go#L340.

I don't understand why this would be necessary.

**Comment by gsalgado** [comment]
@karalabe I think I've addressed all your comments. Can you have another look?
